src/game/logic/object/object.cpp

```diff
@@ -2655,51 +2984,468 @@ void Object::Enter_Group(AIGroup *group)
+
+        if (power && power->Is_Module_For_Power(t)) {
+            return power;
+        }
+    }
+
+    return nullptr;
+}
+
+void Object::Handle_Shroud()
 {
-    return 0;
+    Unlook();
+    Unshroud();
+    Shroud();
+    Look();
 }
 
-void Object::On_Destroy() {}
+void Object::Handle_Value_Map()
+{
+    Remove_Value();
+    Add_Value();
+}
 
-void Object::Restore_Original_Team() {}
+void Object::Handle_Threat_Map()
+{
+    Remove_Threat();
+    Add_Threat();
+}
 
-bool Object::Check_And_Detonate_Booby_Trap(Object *obj)
+void Object::Add_Value()
 {
-    return false;
+    if (m_valueSighting->Is_Invalid()) {
+        if (Get_Controlling_Player()) {
+            if (!Get_Status_Bits().Test(OBJECT_STATUS_UNDER_CONSTRUCTION) && !Is_Effectively_Dead()
+                && Get_Shroud_Clearing_Range() > 0.0f) {
+                m_valueSighting->m_where = *Get_Position();
+                m_valueSighting->m_frame = Get_Template()->Get_Build_Cost();
+                m_valueSighting->m_playerIndex = Get_Controlling_Player()->Get_Player_Mask();
+                m_valueSighting->m_radius = Get_Vision_Range();
+                g_thePartitionManager->Do_Value_Affect(m_valueSighting->m_where.x,
+                    m_valueSighting->m_where.y,
+                    m_valueSighting->m_radius,
+                    m_valueSighting->m_frame,
+                    m_valueSighting->m_playerIndex);
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is adding value, but hasn't removed his previous value.");
+    }
+}
+
+void Object::Remove_Value()
+{
+    if (!m_valueSighting->Is_Invalid()) {
+        g_thePartitionManager->Undo_Value_Affect(m_valueSighting->m_where.x,
+            m_valueSighting->m_where.y,
+            m_valueSighting->m_radius,
+            m_valueSighting->m_frame,
+            m_valueSighting->m_playerIndex);
+        m_valueSighting->Reset();
+    }
+}
+
+void Object::Add_Threat()
+{
+    if (m_threatSighting->Is_Invalid()) {
+        if (Get_Controlling_Player() != nullptr) {
+            if (!Get_Status_Bits().Test(OBJECT_STATUS_UNDER_CONSTRUCTION) && !Is_Effectively_Dead()
+                && Get_Shroud_Clearing_Range() > 0.0f) {
+                m_threatSighting->m_where = *Get_Position();
+                m_threatSighting->m_frame = Get_Template()->Get_Threat_Value();
+                m_threatSighting->m_playerIndex = Get_Controlling_Player()->Get_Player_Mask();
+                m_threatSighting->m_radius = Get_Vision_Range();
+                g_thePartitionManager->Do_Threat_Affect(m_threatSighting->m_where.x,
+                    m_threatSighting->m_where.y,
+                    m_threatSighting->m_radius,
+                    m_threatSighting->m_frame,
+                    m_threatSighting->m_playerIndex);
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is adding threat, but hasn't removed his previous value.");
```

Comment:
Can add object name into this message. Otherwise this assert does not tell much.